### DISCUSSION TITLE
Join Room On Custom Event

### DISCUSSION THREAD

**Discussion by mypeacegarden** [strapi-community/plugin-io#56] [Q&A]
Hi,
I'm trying this plugin. Is there any repo where I can see the implementation?
There are a few things I'm confused about.

I'm using this controller to add new message to my database.
My database has the following collections: `chat-rooms` and `chat-message`
`chat-rooms` contain rooms for duel chats. Suppose user `5` wants to chat with user `7` , I'm searching if there's any room available named `5_7` or `7_5`. If yes, `addNewMessage` will create a new entry and link it with the already existing room (in `chat-rooms`)

````
addNewMessage: async (ctx, next) => {
        const newMessage = await strapi.entityService.create("api::chat-message.chat-message", {
            data: {
                sender_id: {
                    connect: [ctx.state.user.id]
                },
                room_id: {
                    connect: [ctx.request.body.room_id]
                },
                message: ctx.request.body.message
            },
            fields: ['id', 'message'],
            populate: {
                sender_id: {
                    fields: ['id', 'first_name', 'last_name', 'username'],
                    populate: {
                        profile_picture: {
                            fields: ['url']
                        }
                    }
                }
            }
        })
        strapi.$io.emit("api::message.message.newMessage", newMessage)
    },
````

The chat room collection contains `members` and a unique `room_id` . I want to send emits to only those who are the members of the `chat-room` where the message has been added. But I'm not able to let someone join that room This is what I have tried (in the `events` array) so far...

````
{
          "name": "connection",
          "handler": async ({ strapi }, socket) => {
            try {
              let roomId
              let userId = jwtDecode(socket.handshake.auth.token).id

              socket.on("joinSelfRoom", ()=>{
                roomId = userId
                socket.join(roomId)
                socket.emit("roomJoined", {message: `${userId} room joined`}, {room: userId})
              })

              socket.on("joinChatRoom", ()=>{
                roomId = socket.room_id // Want to fetch this room ID from the request
                socket.join(roomId)
                socket.emit("roomJoined", {message: `${userId} room joined`}, {room: userId})
              })
              
            }
            catch (err) {
              console.log(err)
            }
          },
        },
````

````
        {
          "name": "myCustomEvent",
          "handler": async ({ strapi }, socket) => {
            let roomId = JSON.parse(socket).room_id
            strapi.$io.join(roomId) // Not working
            socket.join(roomId) // Not working
          },
        },
````

The client is emitting `myCustomEvent` and the server is able to listen to it

**Comment by ComfortablyCoding** [comment]
Sorry for late response, been a bit busy lately.

What params are you passing when emitting `myCustomEvent` from the client? The socket parameter is not added by default for any event aside from connection (fixed in v2). 

Due to this your emit from the client should look like this
```js
socket.emit("myCustomEvent",socket);
```

In terms of the handler for `myCustomEvent` the 3rd line of code `socket.join(roomId)` should work once the signature is fixed on the client side. Although to get rooms of socket it should be rooms not roomId.

P.S. In terms of examples I do not have any uploaded at the moment but I am working on getting something added that shows all the features.
